"Set Up Secure Signing" Tutorial

mDuo13 opened the pull request:
Includes diagrams and steps for getting yourself an environment where you can sign transactions securely. I figure this will be a key pre-requisite for a bunch of other tutorials.

mDuo13 committed:
Set Up Secure Signing (1st draft)

mellery451 commented on content/tutorials/get-started/set-up-secure-signing.md:
```diff
@@ -0,0 +1,134 @@
+      	port = 5005
+      	ip = 10.1.0.0 # Changed from 127.0.0.1
+      	admin = 10.1.1.2 # Only the client with this IP is treated as admin
+      	protocol = http
```
shouldn't we still recommend `https/wss` for this LAN config?

mDuo13 replied:
Good question. I copied the stanza from the example config and wrote text with the assumption that the communications would be unencrypted, but if https/wss is an option, that would be an improvement.

The question, of course, is the certificates. I don't think having `rippled` use self-signed certs here is a good idea. But setting up rippled to use a proper certificate (assuming that's even possible) from either a well-known CA or an internal CA is probably complicated enough to warrant a separate tutorial.

nbougalis replied:
We should avoid describing this setup _without_ referencing WSS/HTTPS. We should recommend that users install real SSL certificates, issued by a CA and link to Let's Encrypt as one such authority.

mellery451 replied:
ssl certs on a local network seem meh...I mean they are fine, but sure is a lot of work for not much gain. I see value in enabling ssl, just not convinced that certs on the 10. network buy you all that much unless you think something on your LAN can hijack the 10. . That said, I'm not an expert and will defer to Nik et. al.

mDuo13 replied:
I'm kicking this can down the road to a future task. I simplified the instructions for now, and left a draft page to be expanded later.

mellery451 commented on content/tutorials/get-started/set-up-secure-signing.md:
```diff
@@ -0,0 +1,134 @@
+    - [Setup](get-started-with-rippleapi-for-javascript.html)
+    - [API Reference](rippleapi-reference.html)
+- **ripple-libpp for C++**
+    - [Source and Documentation](https://github.com/ripple/ripple-libpp/)
```
I don't think we should recommend ripple-libpp, instead:  https://github.com/ripple/rippled/tree/develop/Builds/linux#signing-library

mellery451 replied:
for that matter, I don't know why we don't just make one or two simple command line utils that sign transactions using this lib and include them with rippled...but we don't for now.

nbougalis replied:
We should make such tools, period. Let's do so for **1.3**.

mDuo13 replied:
Would be nice if they worked in a way that doesn't expose the secret to the process listing and shell history, btw.

mDuo13 committed:
Secure Signing: revs to LAN conf examples & related

nbougalis commented on content/tutorials/get-started/set-up-secure-signing.md:
```diff
@@ -0,0 +1,134 @@
+
+[![Diagram of insecure configurations](img/insecure-signing-options.png)](img/insecure-signing-options.png)
+
+Any configuration in which outside sources may gain access to your secret key is dangerous, and is likely to result in a malicious user stealing all your XRP. Examples of such configurations include ones where you use the [sign method][] of someone else's `rippled` server over the internet, or you transmit your secret key in plain text over the internet to your own server.
```
A malicious user could do more than than steal all your XRP. If you had IOUs they may steal those, etc. Consider rewording to suggest that the malicious user would gain access to the account and be able to authorize the execution of transactions.

nbougalis commented on content/tutorials/get-started/set-up-secure-signing.md:
````diff
@@ -0,0 +1,134 @@
+The following code sample shows how to sign transaction instructions locally with RippleAPI for JavaScript:
+
+```js
+{% include '_code-samples/secure-signing/js/signPayment.js' %}
````
We need to highlight that this snippet is _insecure_ and should not be used as is, in that it includes the key directly in the script file.

mDuo13 replied:
Let's try to figure out a plausible way of demonstrating securely importing the secret instead

nbougalis reviewed (changes requested):
Should we consider highlighting hardware-based signing solutions, like the Ledger? The Trezor Model T also is (theoretically) compatible, thought no wallet capable of leveraging it exists, at this time.

jhaaaa commented on content/tutorials/get-started/set-up-secure-signing.md:
```diff
@@ -0,0 +1,142 @@
+
+    The [example config file](https://github.com/ripple/rippled/blob/8429dd67e60ba360da591bfa905b58a35638fda1/cfg/rippled-example.cfg#L1050-L1073) listens for connections on the local loopback network (127.0.0.1), with JSON-RPC (HTTP) on port 5005 and WebSocket (WS) on port 6006, and treats all connected clients as admin.
+
+    **Caution:** Using the [commandline API](request-formatting.html#commandline-format) for signatures is less secure than [using the Websocket or JSON-RPC APIs](get-started-with-the-rippled-api.html) through non-commandline client. When using the commandline syntax, your secret key may be visible to other users in the system's process listing, and your shell history may save the key in plaintex ...
```
Great point - thank you for this. 

Does this caution also apply to the "Run rippled on the same LAN" option?

mDuo13 replied:
No. (Because if you're connecting on LAN, you're almost certainly not using the commandline interface.)

jhaaaa replied:
Ah, duh. Got it. Thank you!

jhaaaa commented on content/tutorials/get-started/set-up-secure-signing.md:
```diff
@@ -0,0 +1,142 @@
+
+As always, follow industry-standard practices for securing your machines, such as using a firewall, anti-virus, appropriate user permissions, and so on.
+
+**Warning:** This configuration comes with the additional downside that anyone on the LAN can sniff traffic between your machines, potentially gaining access to your secret keys. Do not use this configuration on a network that may have strangers on it. For example, on the LAN at a colocation facility or cloud hosting, other customers may be able to get access to the traffic between your machines. ...
```
possible nit edit? change

>cloud hosting

to

>cloud host

mDuo13 committed:
Secure Signing: rec signing library instead of libpp, etc.

mDuo13 committed:
Set Up Secure Signing: edits per reviews

- Moved the LAN config tutorial to its own (draft) file to be finished later
- Mentioned dedicated signing devices such as the Ledger Nano S
- Improved the JS example's handling of secrets

mDuo13 commented:
- Moved the LAN config tutorial to its own (draft) file to be finished later
- Mentioned dedicated signing devices such as the Ledger Nano S
- Improved the JS example's handling of secrets

mDuo13 committed:
Secure Signing: clarify not just XRP at risk

mDuo13 commented:
(Rebased onto the latest master branch.)

mDuo13 committed:
Secure signing: dedicated hardware, LAN requirement tweaks, etc.

mDuo13 commented:
OK, gonna publish the "Set up Secure Signing" doc as-is and follow up with instructions for setting the LAN config w/ actual certificates in a later task.
